### PR TITLE
Coupling system

### PR DESCRIPTION
**Pull request by hoshiryu**
* **Please check if the PR fulfills these requirements**
- [x] Is the Pull-Request done against the right branch:
  - `master`: for hotfixes not impacting the API
  - `master-v1`: for changes related to the Radium Stable Release V1.
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Be aware that the PR request cannot be accepted if it doesn't pass the Continuous Integration tests.

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
This PR adds `CouplingSystems`, which synchronize sub-systems and dispatch specific calls to them.
This Feature was requested from #324.

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**
For example, the AnimationSystem is now a `CouplingSystem` synchronizing `TimedSystems` through animation-related calls (play/pause/step/cacheFrame/restoreFrame).
Also, for the cacheFrame/restoreFrame calls, a specific directory is created at compile time in the Bundle.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:
IT is preferable to merge #345 before this.

### CODE REVIEW

**Comment by nmellado** [comment]
Is code ready for review ?

File: Plugins/Animation/CMakeLists.txt [hunk 1]
```diff
@@ -70,3 +70,8 @@ target_link_libraries( Animation
                      )
 
 
+add_custom_target( Anim_data_dirs
+    COMMAND ${CMAKE_COMMAND} -E make_directory "${RADIUM_BUNDLE_DIRECTORY}/${CMAKE_BUILD_TYPE}/bin/animFrames"
```
Comment (nmellado) [inline Plugins/Animation/CMakeLists.txt]: The value
```cmake
${RADIUM_BUNDLE_DIRECTORY}/${CMAKE_BUILD_TYPE}/bin
```
is already available as
```cmake
CMAKE_RUNTIME_OUTPUT_DIRECTORY
```
Please replace.

File: Plugins/Animation/src/AnimationComponent.cpp [hunk 2]
```diff
@@ -393,4 +378,55 @@ Scalar AnimationComponent::getTime() const {
     return m_animationTime;
 }
 
+Scalar AnimationComponent::getDuration() const {
+    if ( m_animations.empty() )
+    {
+        return 0;
```
Comment (nmellado) [inline Plugins/Animation/src/AnimationComponent.cpp]:
```c++
return Scalar(0);
```
to avoid implicit cast warnings.

File: Plugins/Animation/src/AnimationComponent.cpp [hunk 3]
```diff
@@ -393,4 +378,55 @@ Scalar AnimationComponent::getTime() const {
...
+    }
+    return m_animations[m_animationID].getDuration();
+}
+
+void AnimationComponent::cacheFrame( int frame ) const {
+    std::ofstream file( "animFrames/"+m_contentName+"_frame"+std::to_string(frame)+".anim",
```
Comment (nmellado) [inline Plugins/Animation/src/AnimationComponent.cpp]: Doesn't Radium has a standard repository for runtime file outputs ?

File: Plugins/Animation/src/AnimationComponent.cpp [hunk 4]
```diff
@@ -393,4 +378,55 @@ Scalar AnimationComponent::getTime() const {
...
+        return;
+    }
+    file.write( (const char*) &m_animationID, sizeof(uint) );
+    file.write( (const char*) &m_animationTimeStep, sizeof(bool) );
+    file.write( (const char*) &m_animationTime, sizeof(Scalar) );
+    file.write( (const char*) &m_speed, sizeof(Scalar) );
+    file.write( (const char*) &m_slowMo, sizeof(bool) );
+    const auto& pose = m_skel.getPose( Ra::Core::Animation::Handle::SpaceType::LOCAL );
+    file.write( (const char*) pose.data(), sizeof(Ra::Core::Transform) * pose.size() );
+    std::cout << "Saving anim data at time: " << m_animationTime << std::endl;
```
Comment (nmellado) [inline Plugins/Animation/src/AnimationComponent.cpp]: Use Radium Log.

File: Plugins/Animation/src/AnimationComponent.cpp [hunk 5]
```diff
@@ -393,4 +378,55 @@ Scalar AnimationComponent::getTime() const {
...
+}
+
+bool AnimationComponent::restoreFrame( int frame )
+{
+    std::ifstream file( "animFrames/"+m_contentName+"_frame"+std::to_string(frame)+".anim",
+                        std::ios::in | std::ios::binary );
```
Comment (nmellado) [inline Plugins/Animation/src/AnimationComponent.cpp]: Same as line 389

File: Plugins/Animation/src/AnimationComponent.hpp [hunk 6]
```diff
@@ -18,39 +18,87 @@ namespace AnimationPlugin {
 
 class SkeletonBoneRenderObject;
 
+/// The AnimationComponent is responsible for the management of the animations of an Entity.
+/// It stores the animation Skeleton and the animation data as well as the animation playing speed.
```
Comment (nmellado) [inline Plugins/Animation/src/AnimationComponent.hpp]: So here, `Animation` means Skeletal Animation (or equivalent).
This cannot be used to animate an object (e.g. camera, position) ?

File: Plugins/Animation/src/AnimationComponent.hpp [hunk 7]
```diff
@@ -18,39 +18,87 @@ namespace AnimationPlugin {
...
+
+    /// Create the skeleton from the given data.
+    /// @param data: the skeleton's joint transform hierarchy.
+    /// @param duplicateTable: the map from duplicate to single vertices.
+    /// @param the number of single vertices.
+    // FIXME: the 2 last parameters are needed only for the genereation of
+    //        the skinning weights matrix.
+    void handleSkeletonLoading( const Ra::Asset::HandleData* data,
+                                const std::vector<Ra::Core::Index>& duplicateTable,
+                                uint nbMeshVertices );
```
Comment (nmellado) [inline Plugins/Animation/src/AnimationComponent.hpp]: It is not clear to me why we still have the duplicate table.
Isn't the attribute system supposed to fix this ?

File: Plugins/Animation/src/AnimationComponent.hpp [hunk 8]
```diff
@@ -18,39 +18,87 @@ namespace AnimationPlugin {
...
-    void handleSkeletonLoading( const Ra::Asset::HandleData* data,
-                                const std::vector<Ra::Core::Index>& duplicateTable,
-                                uint nbMeshVertices );
-    void handleAnimationLoading( const std::vector<Ra::Asset::AnimationData*> data );
+    /// @returns the duration of the current animation.
+    Scalar getDuration() const;
+
+    // Skeleton display
+    /// Turns xray display on/off for the skeleton bones.
+    void setXray( bool on ) const;
```
Comment (nmellado) [inline Plugins/Animation/src/AnimationComponent.hpp]: Is this really related to the animation ? Shouldn't this be a property of the skeleton ?

File: Plugins/Animation/src/AnimationComponent.hpp [hunk 9]
```diff
@@ -18,39 +18,87 @@ namespace AnimationPlugin {
...
+
+    /// Toggle skeleton bones display.
+    void toggleSkeleton( const bool status );
```
Comment (nmellado) [inline Plugins/Animation/src/AnimationComponent.hpp]: Same as above

File: Plugins/Animation/src/AnimationComponent.hpp [hunk 10]
```diff
@@ -18,39 +18,87 @@ namespace AnimationPlugin {
...
+
+    /// @returns the index of the skeleton bone associated to the RenderObject with index \p index.
+    uint getBoneIdx( Ra::Core::Index index ) const;
```
Comment (nmellado) [inline Plugins/Animation/src/AnimationComponent.hpp]: Same as above

File: Plugins/Animation/src/AnimationComponent.hpp [hunk 11]
```diff
@@ -63,54 +111,86 @@ class ANIM_PLUGIN_API AnimationComponent : public Ra::Engine::Component {
     virtual void setTransform( Ra::Core::Index roIdx,
                                const Ra::Core::Transform& transform ) override;
 
-  public:
+private:
+    // Internal function to create the skinning weights.
+    void createWeightMatrix( const Ra::Asset::HandleData* data,
+                             const std::map<uint, uint>& indexTable,
+                             const std::vector<Ra::Core::Index>& duplicateTable,
+                             uint nbMeshVertices );
```
Comment (nmellado) [inline Plugins/Animation/src/AnimationComponent.hpp]: same as above.

File: Plugins/Animation/src/AnimationComponent.hpp (hunk 6 above)
Comment (nmellado) [reply Plugins/Animation/src/AnimationComponent.hpp]: Update: after reading the rest of the code.
The animation component seems to be in charge of many other things, like creating the skeleton bones for display.
Please clearly state the role of this object.

File: Plugins/Animation/src/AnimationSystem.cpp [hunk 12]
```diff
@@ -118,7 +152,7 @@ void AnimationSystem::handleAssetLoading( Ra::Engine::Entity* entity,
         }
 
         // FIXME(Charly): Certainly not the best way to do this
```
Comment (nmellado) [inline Plugins/Animation/src/AnimationSystem.cpp]: What do we do with this FIXME(Charly) ?

File: src/Engine/System/CouplingSystem.hpp [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    /// Saves all the state data related to the \p frameID -th frame into a cache file.
+    virtual void cacheFrame( int frameID ) const = 0;
+
+    /// Restores the state data related to the \p frameID -th frame from the cache file.
+    /// \returns true if the frame has been successfully restored, false otherwise.
+    virtual bool restoreFrame( int frameID ) = 0;
+};
+
+/// Coupling Systems are responsible for transmitting calls to a bunch of other Systems.
+template <typename CSystem>
```
Comment (nmellado) [inline src/Engine/System/CouplingSystem.hpp]: Is `CSystem` always inheriting `CoupledSystem` ?
If yes, you don't need the template parameter.
If no, why do you need a template parameter, and why not using `System *` directly ?

File: Plugins/Animation/src/AnimationComponent.hpp (hunk 7 above)
Comment (hoshiryu) [reply Plugins/Animation/src/AnimationComponent.hpp]: Well, for now, the skinnnig weights are not stored as attributes but as a matrix.
And when creating this matrix, which has a row per duplicated position, we must know how to map the loaded indices (in GeometryData) to the stored indices (in TriangleMesh).

File: Plugins/Animation/src/AnimationComponent.hpp (hunk 6 above)
Comment (hoshiryu) [reply Plugins/Animation/src/AnimationComponent.hpp]: I would say, without taking a deep look at it, it should be renamed AnimationSkeletonComponent, though I have to investigate about its use for other animation purposes (e.g. camera, position, ...)

File: Plugins/Animation/src/AnimationSystem.cpp (hunk 12 above)
Comment (hoshiryu) [reply Plugins/Animation/src/AnimationSystem.cpp]: I don't know either, will take a look at when it was posted

File: src/Engine/System/CouplingSystem.hpp (hunk 13 above)
Comment (hoshiryu) [reply src/Engine/System/CouplingSystem.hpp]: `CSystem` stands for `CoupledSystem` and which might provide the members required by the `CouplingSystem` managing it.
Having only `System*`, how can the `CouplingSystem` call the required members?

Comment (nmellado) [reply src/Engine/System/CouplingSystem.hpp]: So, just storing pointers of CoupledSystem should do the job, no ?

File: Plugins/Animation/CMakeLists.txt (hunk 1 above)
Comment (hoshiryu) [reply Plugins/Animation/CMakeLists.txt]: well, this is set to `RADIUM_PLUGIN_OUTPUT_PATH` in each plugin's cmake... is this something to change?

File: Plugins/Animation/src/AnimationComponent.cpp (hunk 3 above)
Comment (hoshiryu) [reply Plugins/Animation/src/AnimationComponent.cpp]: Actually, these runtime outputs are not relevant only for the current session, it's more like added data to the scene.
I will change that and have the user specify where he wants to save these data (better be along the scene file)

File: Plugins/Animation/src/AnimationComponent.hpp (hunk 6 above)
Comment (hoshiryu) [reply Plugins/Animation/src/AnimationComponent.hpp]: For now, the `AnimationPlugin` only deals with skeleton-based character animation.
Two choices: 
 - just specify this in the doc for now, and improve the animation management in another PR.
 - include it in this PR, but then postpone its review/acceptance to let other PRs pass.

Regarding the `AnimationComponent`, I don't really know how to improve it. A possible solution would be the following.
Let the `AnimationComponent` only store the `Core::Animation`s and current animation id and time. It will also expose the current animation and pose.
Then there could be an `AnimationSkeletonComponent` which would only store the `Core::Animation::Skeleton` and a map from bone id to transform storage index in the animation pose. It would also be responsible for updating the skeleton pose, through Component IO, and display it.

But I'm not sure about it, and if we want to handle other animation purposes, we need a deep reflection on how to deal with such cases, i.e. how to update the animated data.
Let's say we have a cube, with keyframed transforms. Then the `AnimationComponent` might store all the transforms, but how to / who should change the cube's transform when playing the animation? And what about animated cameras?
We need a mechanism to bind the animation transform to its use.

File: Plugins/Animation/CMakeLists.txt (hunk 1 above)
Comment (nmellado) [reply Plugins/Animation/CMakeLists.txt]: Interesting indeed.
Yes this needs to be fixed.
The variables (`CMAKE_RUNTIME_OUTPUT_DIRECTORY`,...) shouldn't be modified in each Plugin CMakeList.
If needed, directly use `RADIUM_PLUGIN_OUTPUT_PATH` and directly set the required targets properties.

Comment (hoshiryu) [reply Plugins/Animation/CMakeLists.txt]: shall we do this in this PR? I remember reordering such things within the test_install* branches about the installation procedure of Radium.

Comment (nmellado) [reply Plugins/Animation/CMakeLists.txt]: Ok, let's keep it for another PR, when we will rework the build system (install)

File: Plugins/Animation/src/AnimationComponent.hpp (hunk 7 above)
Comment (nmellado) [reply Plugins/Animation/src/AnimationComponent.hpp]: Please fix this in a dedicated PR.

**Comment by hoshiryu** [comment]
Ready for review/acceptance
